### ISSUE TITLE
Feature Request: Webhooks

### ISSUE THREAD

**Issue by driverpt** [ory/kratos#271]
**Is your feature request related to a problem? Please describe.**
In a Microservices based environment, when a new user is registered and/or activated, we have the need to provision resources in other services for the newly created/activated user.

**Describe the solution you'd like**
Everytime a user is created and/or activated, we'd like support to have Webhooks to other services with a well-known JSON Structure.

e.g.:
User A registers in Kratos. (Webhook triggered for Services A which posts in a Message Queue the event). Service D, E and F, provision resources for this user (e.g.: Storage, Security and ACL).
User A is activated in Kratos (Email Activation). All Resources are ready to go.

**Describe alternatives you've considered**

I've considered creating a microservice and have and endpoint for everytime a user logs in, but seems a bit overkill to have an endpoint +1 redirect just to do these checks and provisioning when it can be done via WebHook

**Additional context**

None

**Comment by aeneasr** [comment]
Yes, that was on the roadmap :) Triaging.

This is btw also supported with after/before hooks - all we need is a JSON RPC handler there, contributions welcome!

**Comment by zepatrik** [comment]
I think we should only send the identity's id though so you can gather any data yourself.

**Comment by aeneasr** [comment]
It's also important to note that there are two types of webhooks:

- Broadcast which would just send out an event such as "user.signedup" with some identity data
- RPC which would send and also receive data to modify the identity for example

The second is pretty difficult to achieve as it also has a certain toll on request latency, the first one is pretty easy to achieve.

**Comment by driverpt** [comment]
OK. I'll give it a look and try to make a PR. Is there any internal message queue in the code for these events so we can just create Event Handlers?

**Comment by zepatrik** [comment]
Nope, the only messaging we have right now is the email verification. Webhooks should be implemented like the other "hooks" (sorry for the bad wording here). That way you can easily configure them as pre/post registration/login/...
Have a look in the `selfservice/hook` module.

**Comment by aeneasr** [comment]
Here's the list of relevant code lines:

- Hook names: https://github.com/ory/kratos/blob/master/selfservice/hook/hooks.go#L7
- Hook interface for post/pre sign up: https://github.com/ory/kratos/blob/master/selfservice/flow/registration/hook.go#L17-L22
- Hook interface for post/pre login: https://github.com/ory/kratos/blob/master/selfservice/flow/login/hook.go#L12-L17
- Example hook for both post login and sign up: https://github.com/ory/kratos/blob/master/selfservice/hook/redirector.go#L38-L44
- Registering hook in the central "service manager" / "service locator":  https://github.com/ory/kratos/blob/master/driver/registry_default_hooks.go

The registry (`RegistryDefault`) contains all the services that run in Kratos - e.g. the service that persists data, the service that handles login, the handler ("controller" in classical MVC) that handles REST requests for getting identities, ...

Hope this helps

**Comment by aeneasr** [comment]
One more thing:

Hooks are being called from each individual strategy (e.g. username + password, oidc, ...):

- https://github.com/ory/kratos/blob/master/selfservice/strategy/password/registration.go#L165-L174
- which jumps call to here: https://github.com/ory/kratos/blob/230ab2d83f4d187f410e267c6d68554e82514948/selfservice/flow/login/hook.go#L42
- which in turn then uses the registry to get the right hook and execute the logic

**Comment by driverpt** [comment]
Hey,

I'm not an expert in go and i would rather that this is written by someone more experient than me, but i did a small patch:

[webhook_poc.patch.zip](https://github.com/ory/kratos/files/4303931/webhook_poc.patch.zip)

**Comment by jakhog** [comment]
What kind if delivery guarantees are you envisioning for these webhooks? If you are aiming for best effort, then spinning up a goroutine doing the work should be quite simple as mentioned. If you are imagining a higher QoS, with persisted messages, then there are quickly some more difficult things with scaling etc. In that case, maybe an outgoing-message-delivery system could be it's own thing outside Kratos, that could also be used in the other Ory projects?

I for one would love push-notifications from the Ory components when things happen, either webhooks or possibly through some other pub-sub system.

**Comment by zepatrik** [comment]
I would say we should start off with the "best effort" stategy and add a message delivery system later on. As we're currently still in alpha it might be more valuable to get it running fast and then incorporate the feedback from the community instead of going with a high QoS from the beginning.

**Comment by eshepelyuk** [comment]
Isn't it related to #1030 and #776 ?

**Comment by dadrus** [comment]
Hello @aeneasr,

We (I'm talking about my team which tries to adopt kratos) have implemented a possible approach for web hooks, which we would like to discuss with you, respectively the community. This approach does however not follow the hooks concept kratos currently implements. The reasoning behind is that not every self service flow has a possibility to add a strategy (like link, password, oidc, whatever might come in the future). One example is the verification flow. To be more specific, the idea is to have the web hooks registered like depicted in the following kratos.yml snippet:

```
....
   registration:
      lifespan: 10m
      ui_url: http://127.0.0.1:8080/my_registration_ep
      after:
        default_browser_return_url: http://127.0.0.1:8080/my_activation_pending_ep
        web_hooks:
          - http://127.0.0.1:8080/do_whatever_is_required_1
          - http://127.0.0.1:8080/do_whatever_is_required_2

    verification:
      enabled: true
      ui_url: http://127.0.0.1:8080/my_activate_account
      after:
        default_browser_return_url: http://127.0.0.1:8080/my_account_activated
        web_hooks:
          - http://127.0.0.1:8080/do_whatever_is_required_3
....
```
As of today the implementation is straight forward in the sense, that 
* as a method HTTP POST is used
* as payload the identity object of the registered user is used
* the web hooks are supported for registration and verification only

I can make a pull request to be considered a WIP, where you can have more insights on the code and which, if the approach is ok, can be extended to cover possible further demand (whatever it is).

**Comment by aeneasr** [comment]
Thank you! I think it would be easier to add hooks to verification instead of introducing a completely new config key / methodology. I also think we would need to change this a bit and use things like e.g. JSONNet for maximum usability:

```
selfservice:
  flows:
    login:
      after:
        password:
          - hook: json-rpc
            config:
              method: POST # PUT, PATCH, DELETE
              url: https://...
              body: path/to/my/jsonnet
              authorization: ...
```

The jsonnet gets some context (e.g. the identity, the flow, request headers, ...) and the result of the jsonnet is sent using the method. If `method: GET`, the body portion would be ignored.

Last but not least this would need some type of authentication. Webhooks should be able to certify incoming requests. The easiest would probably be to specify that in the authorization header (e.g. a pre-shared key or something).

**Comment by dadrus** [comment]
Thank you for the feedback!. We can for sure adjust the current implementation, but then there is a "main" question: which flows should be supported? As I've written in my previous comment, we just added "support" for registration and verification flows. The latter one does however not have any "object", like password or oidc, where a hook could be configured

So if the idea is to have the web hooks supported by all flows, based on your answer, it is clear, how the configuration could look like for registration, login and settings. But how should it be then configured for logout, verification, recovery and error flows? In our case we urgently need a web hook for verification.

Thank you in advance.

**Comment by abador** [comment]
Hi @dadrus and @aeneasr I can help out with this. We're trying to integrate Kratos with some of our services. @dadrus   I can take a look into your WIP (it would be also easier to discuss this).

**Comment by dadrus** [comment]
Based on the comments from @aeneasr we've updated the implementation to have the configuration for the web-hooks to look like

```
- hook: json-rpc
  config:
    method: POST # PUT, PATCH, DELETE
    url: https://...
    body: path/to/my/jsonnet
    authorization: ...
```
We're now working on adding support for JSONNet. When this is done, I'll create a pull request.

Nevertheless it would be great to have the answers for the questions, I've asked in my previous comment.

**Comment by mboutet** [comment]
Using the proposed schema to define the hooks, would it be possible later on to introduce something akin to `failurePolicy` of [Kubernetes webhooks](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#mutatingwebhook-v1-admissionregistration-k8s-io)? This would be very useful for webhooks that absolutely need to be successful for ensuring the consistency of the whole system in which Kratos is integrated.

In other words, is the proposed implementation suitable for adding such a functionally in a backward-compatible way in the future?

**Comment by aeneasr** [comment]
Not sure if we want to be held responsible here for consistency guarantees. That sounds really difficult to achieve. We can definitely add some retry mechanisms. For consistency I would fall back to some querying mechanism or something else.

Hooks already support per-strategy and "global" runs @dadrus so not sure if I understand your question?

I think we need to add hook processing to all other flows as well.

**Comment by aeneasr** [comment]
In the PR it would be better to add the new RPC hook first, and then another PR to solve the other things

**Comment by mboutet** [comment]
@aeneasr, what I mean by consistency is ensuring that a given operation in Kratos (e.g. user registration) is only successful if all the webhooks with `failurePolicy=Fail` are successful. For instance, an API needs to know which subjects exist. In order to do that, it needs to be notified of new registration. If for some reason the webhook fails and Kratos just ignore this failure, then the user would exist in Kratos, but not in this other API thus leading to an inconsistent state.

So, in my opinion, the only think Kratos should implement in regards to this is return an error if any required webhooks fail.

**Comment by dadrus** [comment]
@aeneasr:

> Hooks already support per-strategy and "global" runs @dadrus so not sure if I understand your question?

Given the following example

```
selfservice:
  flows:
    verification:
      after:
        all: <--------------------------------------------------
          - hook: json-rpc
            config:
              method: POST # PUT, PATCH, DELETE
              url: https://...
              body: path/to/my/jsonnet
              authorization: ...
```

(1) What should be the name for the "global" hooks? "all", "global", something else? I was unable to find anything related to it in the code.

> I think we need to add hook processing to all other flows as well.
This answers most of the other questions ;)
In addition to the question from above another one is however still open:
(2) What would be the expected behavior if web hooks are configured for a specific strategy and on the "global" level? With other words, does a more specific web hook overwrite the one on a global level if the configuration is the same? In my opinion - no.

Thank you in advance.

**Comment by dadrus** [comment]
One additional thing: Let us better use "web-hook" as hook type and not "json-rpc". For me "json-rpc" implies, kratos would support the [JSON RPC](https://www.jsonrpc.org/specification) specification. I don't thing we want that support.

**Comment by aeneasr** [comment]
> @aeneasr, what I mean by consistency is ensuring that a given operation in Kratos (e.g. user registration) is only successful if all the webhooks with failurePolicy=Fail are successful. For instance, an API needs to know which subjects exist. In order to do that, it needs to be notified of new registration. If for some reason the webhook fails and Kratos just ignore this failure, then the user would exist in Kratos, but not in this other API thus leading to an inconsistent state.
>
> So, in my opinion, the only think Kratos should implement in regards to this is return an error if any required webhooks fail.

Even that might be difficult as we would have to remove all the records related to the flow again. It's theoretically possible - just not easy to implement. It might be a better idea to have a Kafka adapter in place or something that ensures delivery!

> Hooks already support per-strategy and "global" runs @dadrus so not sure if I understand your question?

It looks like this was possible before but we removed it. I would prefer a syntax like this:

```
selfservice:
  flows:
    verification:
      password:
        after: # only for password
          - hook: foo
      after: # global for verification
        - hook: bar
```

WDYT?

**Comment by mboutet** [comment]
> It might be a better idea to have a Kafka adapter in place or something that ensures delivery!

Not sure if forcing the integrator to deploy and/or operates Kafka is a sensible idea. Also, what about if Kafka is somehow unavailable/unreachable? You got the same problem as a failed webhook. Besides, I think that from an integrator perspective, webhooks offer the most flexible solution because the webhook-receiving service can do whatever it wants: Kafka, RabbitMQ, Google Pub/Sub, Azure service bus, gRPC calls to other downstream services, etc.

I'll open an issue once the initial webhook implementation is released to discuss this feature.

Thanks!

**Comment by dadrus** [comment]
@aeneasr :

**Comment by abador** [comment]
> (2) What would be the expected behavior if web hooks are configured for a specific strategy and on the "global" level? With other words, does a more specific web hook overwrite the one on a global level if the configuration is the same? In my opinion - no.

In a long run overrides are tricky at best i would rather go with @aeneasr 's suggestion with running both global and flow hooks

> > It might be a better idea to have a Kafka adapter in place or something that ensures delivery!
> 
> Not sure if forcing the integrator to deploy and/or operates Kafka is a sensible idea. Also, what about if Kafka is somehow unavailable/unreachable? You got the same problem as a failed webhook. Besides, I think that from an integrator perspective, webhooks offer the most flexible solution because the webhook-receiving service can do whatever it wants: Kafka, RabbitMQ, Google Pub/Sub, Azure service bus, gRPC calls to other downstream services, etc.
> 
> I'll open an issue once the initial webhook implementation is released to discuss this feature.
> 
> Thanks!

In my opinion this should be done as lightweight as possible. Any MQ integration can be done using web hooks, and I would really like for it to be as independent as possible given the number of solutions on the market. 
To provide consistency and a valid retry mechanism maybe we should consider storing failed hook requests in the database and provide some kind of a retry in the webhook queue? In such cases a replication lag should always be taken into consideration so it looks like valid way to go without rollbacks.

**Comment by manjeshpv** [comment]
I suggest failure logs will be written to log file or error monitoring systems

@aeneasr keeping minimal/simple/lightweight webhook implementation in the first phase  

Similar implementations: https://docs.min.io/docs/minio-bucket-notification-guide.html

platforms ensure reliable delivery https://hookdeck.com/
